src/newgrf.cpp

```diff
@@ -9478,21 +9483,30 @@ static void FinaliseIndustriesArray()
 		}
 		if (!indsp.enabled) {
 			indsp.name = STR_NEWGRF_INVALID_INDUSTRYTYPE;
+			continue;
 		}
 
 		/* Apply default cargo translation map for unset cargo slots */
-		for (uint i = 0; i < lengthof(indsp.produced_cargo); ++i) {
-			if (!IsValidCargoID(indsp.produced_cargo[i])) indsp.produced_cargo[i] = GetCargoIDByLabel(GetActiveCargoLabel(indsp.produced_cargo_label[i]));
+		for (auto it = std::begin(indsp.produced_cargo); it != std::end(indsp.produced_cargo); ++it) {
+			if (!IsValidCargoID(*it)) *it = GetCargoIDByLabel(GetActiveCargoLabel(indsp.produced_cargo_label[std::distance(std::begin(indsp.produced_cargo), it)]));
+			/* If cargo is produced already, disable this slot. */
+			if (IsValidCargoID(*it) && std::find(std::begin(indsp.produced_cargo), it, *it) != it) *it = INVALID_CARGO;
```

Comment:
Should it throw a warning to the user? On `newgrf, 1` or something? No clue if that is useful at all.

Reply:
I don't think the line number matters, if you specify which industry has the problem. I think a pop-up warning is a good idea to help GRF authors catch their mistakes.